components/lib/push-amazon/src/main/java/mozilla/components/lib/push/amazon/AbstractAmazonPushService.kt

```diff
@@ -0,0 +1,81 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ *  License, v. 2.0. If a copy of the MPL was not distributed with this
+ *  file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+package mozilla.components.lib.push.amazon
+
+import android.content.Context
+import android.content.Intent
+import android.os.Bundle
+import com.amazon.device.messaging.ADM
+import com.amazon.device.messaging.ADMMessageHandlerBase
+import mozilla.components.concept.push.EncryptedPushMessage
+import mozilla.components.concept.push.PushError
+import mozilla.components.concept.push.PushProcessor
+import mozilla.components.concept.push.PushService
+import androidx.annotation.VisibleForTesting
+
+/**
+ * An Amazon Cloud Messaging implementation of the [PushService] for Android devices that support Google Play Services.
+ * [ADMMessageHandlerBase] requires a redundant constructor parameter.
+ */
+abstract class AbstractAmazonPushService : ADMMessageHandlerBase("AbstractAmazonPushService"), PushService {
+    private lateinit var adm: ADM
+
+    override fun start(context: Context) {
+        adm = ADM(context)
+        if (adm.registrationId == null) {
+            adm.startRegister()
+        }
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PROTECTED)
+    public override fun onRegistered(registrationId: String?) {
+        registrationId?.let { PushProcessor.requireInstance.onNewToken(it) }
+    }
+
+    override fun onRegistrationError(errorId: String?) {
+        errorId?.let { PushProcessor.requireInstance.onError(
+                PushError.MalformedMessage("registration failed: $errorId")
+            )
+        }
+    }
+
+    override fun onUnregistered(registrationId: String?) {
```

Comment:
@jonalmeida Do we have anything in `PushProcessor` or `AutopushFeature` to handle unregistration?

Reply:
From the API docs:
> Called on successful unregistration. This method may be called in response to your app calling startUnregister() or if ADM has unregistered the app for some reason (**typically because the device has lost its association with a user's Amazon account**).

This is interesting! While there isn't anything to handle the unregistration, we could surface the error with `PushProcessor.onError(PushError.Registration("account no longer associated with device"))`.

However, [there are situations](https://github.com/mozilla-mobile/android-components/issues/3859) where we need to force an unregistration, so we should make sure to only propagate the error in that case where we didn't explicitly call `startUnregister`.